src/conn/conn_compact.c

```diff
@@ -18,13 +20,233 @@ __compact_server_run_chk(WT_SESSION_IMPL *session)
+        return (false);
+    }
+
+    return (true);
+}
+
+/*
+ * __background_compact_start --
+ *     Pre-fill compact related statistics for the given file.
+ */
+static int
+__background_compact_start(
+  WT_SESSION_IMPL *session, const char *uri, WT_BACKGROUND_COMPACT_STAT **compact_stat)
+{
+    WT_BACKGROUND_COMPACT_STAT *temp_compact_stat;
+    WT_BM *bm;
+    WT_DECL_RET;
+
+    bm = S2BT(session)->bm;
+
+    temp_compact_stat = __background_compact_get_stat(session, uri);
+
+    /* If the table is not in the list, allocate a new entry and insert it. */
+    if (temp_compact_stat == NULL) {
+        WT_ERR(__wt_calloc_one(session, &temp_compact_stat));
+        WT_ERR(__wt_strdup(session, uri, &temp_compact_stat->uri));
+        __background_compact_list_insert(session, temp_compact_stat);
+    }
+
+    /* Fill starting information prior to running compaction. */
+    WT_ERR(bm->size(bm, session, &temp_compact_stat->start_size));
+    temp_compact_stat->prev_compact_time = __wt_clock(session);
+
+    *compact_stat = temp_compact_stat;
+
+    return (0);
+
+err:
+    __wt_free(session, temp_compact_stat);
+
+    return (ret);
+}
+
+/*
+ * __background_compact_end --
+ *     Fill resulting compact statistics in the background compact tracking list for a given file.
+ */
+static int
+__background_compact_end(WT_SESSION_IMPL *session, WT_BACKGROUND_COMPACT_STAT *compact_stat)
+{
+    WT_BM *bm;
+    WT_CONNECTION_IMPL *conn;
+    wt_off_t bytes_recovered;
+
+    bm = S2BT(session)->bm;
+    conn = S2C(session);
+
+    WT_RET(bm->size(bm, session, &compact_stat->end_size));
+    compact_stat->bytes_rewritten = bm->block->compact_bytes_rewritten;
+    bytes_recovered = compact_stat->start_size - compact_stat->end_size;
+
+    /*
+     * If the file failed to decrease in size, mark as an unsuccessful attempt. It's possible for
+     * compaction to do work (rewriting bytes) while other operations cause the file to increase in
+     * size.
+     */
+    if (bytes_recovered <= 0) {
+        compact_stat->consecutive_unsuccessful_attempts++;
+        compact_stat->prev_compact_success = false;
+    } else {
+        compact_stat->consecutive_unsuccessful_attempts = 0;
+        conn->background_compact.files_compacted++;
+        compact_stat->prev_compact_success = true;
+
+        /*
+         * Update the moving average of bytes rewritten across each file compact attempt. A
+         * weighting of 10% means that we are effectively considering the last 10 attempts in the
+         * average.
+         */
+        conn->background_compact.bytes_rewritten_ema =
+          (uint64_t)(0.1 * bm->block->compact_bytes_rewritten +
+            0.9 * conn->background_compact.bytes_rewritten_ema);
+    }
+
+    return (0);
+}
+
+/*
+ * __background_compact_list_cleanup --
+ *     Free all or any entry that has not been updated for more than a day in the background compact
+ *     tracking list.
+ */
+static void
+__background_compact_list_cleanup(
```

Comment:
General question about statistics maintenance:
 - Compaction statistics are created for eligible uri-a
 - uri-a is dropped (deleted).
 - In < 24hrs a new uri-a is created.
 - If bg compact decides to compact uri-a it will use the previous statistics.

Reply:
This is such a good observation and an interesting problem. We have unique IDs for checkpoints, do we have something similar for dhandles/URIs @agorrod?

Reply:
Interesting. There is a btree ID (https://github.com/wiredtiger/wiredtiger/blame/develop/src/include/btree.h#L113) that uniquely identifies a table, you could potentially use that, though, to be honest I'm a little uncomfortable with how many places seem to be using that (it is there for WAL implementation).

I also suspect that it will need to be used in conjunction with the URI in many places, since it isn't obviously available (though maybe it is, the metadata cursor might be able to retrieve the btree ID?)

Reply:
That's an excellent point @marcbutler and thanks for pointing out the btree ID @agorrod. That file ID is present in the configuration string as file metadata. The metadata cursor can retrieve this configuration string and we can then parse it to retrieve the ID field. As you mentioned, we can use that in conjunction with the URI but we only need to do that when inserting and retrieving entries from the list so I don't think that's too complicated.

Reply:
I've added the handling of this case in https://github.com/wiredtiger/wiredtiger/pull/9541/commits/138801431ace7c1d77267584a57f12d92db03d2f.

Reply:
As I'm not familiar with the intricacies of the data structures: but from what I can discern the code is now looks to prevent the problem. So I'm going to ride into the sunset...

@agorrod Out of curiosity (I'd never of survived kittenhood) what is "WAL implementation" in this context.